### DISCUSSION TITLE
another seed abomination issue

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#219] [Design & Development]
Worked around the bug to make some progress on the short course tutorials.  I've developed a full receiver function deconvolution script with a fairly complex string of algorithms.  Serial runs on this little desktop are running about 1 seismogram object per s.  That is about 2 orders of magnitude larger than the one we used for the paper.   Might be a good test run for a typical algorithm.

Now, the bad news is when I working through this I ran into a minor issue with the default schema that we may want to debate.   The problem is the quantities lat, lon, elev are required to be loaded (as they should) but seed network and station names are not being loaded in normalization because they are marked in the  yaml file as "normal" while things like lat, lon, and elev are listed as "required".  I totally agree with that, but using it reveals a hole in the our reader api.   

The problem is there is no way to cleanly keep the seed codes with the data.   The save_data method called on Seismogram objects contain net, sta, or loc will generate attributes (by default) with the reaonlyerror tags:  READONLYERROR_sta, READONLYERROR_net, and READONLYERROR_loc.   (chan too, but that is irrelevant for Seismogram, although the same will happen for saving TimeSeries data).   Here is an example from my working notebook:
```
_id 60dccbeeeef8d7cf53ba4aab
_fdsnws_dataselect_url http://service.iris.edu/fdsnws/dataselect/1/query
_format MSEED
calib 1.0
chan BHZ
channel_edepth 0.0
channel_elev 0.062
channel_endtime 1340387999.0
channel_id 60dcc52aeef8d7cf53ba368d
channel_lat 33.257702
channel_lon -92.803001
channel_starttime 1299715200.0
delta 0.025
endtime 1302188764.100001
gridfs_id 60dccbeeeef8d7cf53ba4aac
mseed {'dataquality': 'M', 'number_of_records': 17, 'encoding': 'STEIM2', 'byteorder': '>', 'record_length': 4096, 'filesize': 100441600}
READONLYERROR_net TA
npts 59081
processing ['ObsPy 1.2.2: trim(endtime=UTCDateTime(2011, 4, 7, 15, 6, 4, 100000)::fill_value=None::nearest_sample=True::pad=False::starttime=UTCDateTime(2011, 4, 7, 14, 41, 27, 100000))']
sampling_rate 40.0
site_id 60dcc52aeef8d7cf53ba368b
READONLYERROR_source_depth 53.2
source_id 60dcc513eef8d7cf53ba3177
READONLYERROR_source_lat 38.2513
READONLYERROR_source_lon 141.7296
READONLYERROR_source_magnitude 7.1
READONLYERROR_source_time 1302186764.1
READONLYERROR_sta Z41A
starttime 1302187287.100001
storage_mode gridfs
time_standard UTC
utc_convertible True
history_object_id 59e8a7ad-a3b7-4ac4-8a45-86398b5be33c
elog_id 60dccbeeeef8d7cf53ba4aaa
data_tag rawdata
```
Note the clean method can be used to clear this and all the other debris, BUT when you do that you leave only site_id.   That is, in fact, what we should want.  The problem then comes is is if you try to read the data back with a simple line like:
```
d=dbh.read_data(doc,normalize=['site','source'])
```
there is currently NO WAY to fetch the net or sta fields without doing a bunch of indendent MongoDB queries.

With that long description here is the point:  we need a new parameter for the read_data method to handle data loaded via the normalization method for site, channel, and source.   Specifically, we need a clean way to tell the reader we want to add list x of attributes in addition to the required ones.   A simple solution would be to instead of sending a list of collection names for normalization you could send a dict with a form like this:
```
{ 'site' : [ 'net', 'sta'] , 'source' : [] }
```
where this would tell the reader that for site it should append net and sta to the list of metadata to load and for source because it is empty add no new fields.    If we used this, I think the reader would need to test for the type of normalize argument. If it is a list behave as it does not but if it is a dict add the logic to append to the list of attributes to normalize.

There are likely better ways to do this.  My suggestion is what came to me as a rookie python programmer.  Other ideas?

**Comment by wangyinz** [comment]
hmmm.... I don't understand it. I thought when `normalize` is given, everything from that collection should be read. Is that not the case? @haruming

**Reply by wangyinz** [reply]
e.g, this set of test should cover exactly the issue, right?
https://github.com/mspass-team/mspass/blob/82d65829fc6b774ac529619760a05e76a446bdae/python/tests/db/test_db.py#L441-L446

**Reply by wangyinz** [reply]
I ran this test in the debugger and can see that the `promiscuous_seis2` does have 'net' and 'sta' defined.

**Reply by wangyinz** [reply]
nonetheless, the proposed feature is very useful no matter what.

**Comment by haruming** [comment]
You are right if we save the metadata with attributes marked as read-only, it would be saved as READONLYERROR_xxx like this. However, when we call read_data if we provide a list of normalization collection names, it would load all the attributes of the normalized collection according to the normalized_id in the object_doc in MongoDB. It simply ignores the READONLYERROR_xxx in the object_doc unless we specify the reader with promiscuous mode.

Therefore, as long as we have the site_id, source_id and channel_id associated with the object_doc, we could retrieve all the normalized attributes and load them into a mspass_object. Even if we clean the object_doc, it still contains a site_id, and the reader could load net and sta from my understanding unless we can't find the corresponding record in the site collection with that site_id.

Well, I agree with what Ian said it is a useful feature if we could provide something like a dict indicating what specific attributes we would like to load instead of loading all the attributes within that collection.
